### PR TITLE
Jennyf/up

### PR DESCRIPTION
**Pull request by jennyf19**
I have not completed the required manual/automated tests for this branch, but wanted to get a PR out for review. I will update the tests as I complete them.

Only changes starting with [commit 05541d8](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/commit/05541d804a45c6b00e2261467806d4ec5e03b6f5) need review, as all previous commits have already been reviewed.

**Testing completed:**
> ADALDesktopTestApp (u/p) - manual
> MSAL DesktopTestApp (u/p) - manual

### CODE REVIEW

File: msal/samples/desktop/SampleApp/MainForm.cs [hunk 1]
```diff
@@ -52,31 +56,70 @@ public MainForm()
...
+
         private async void TabControl1_SelectedIndexChanged(object sender, EventArgs e)
         {
-            if((sender as TabControl).TabIndex == 1)
+            if ((sender as TabControl).TabIndex == 1)
             {
-                string token = await _msalHelper.GetTokenForCurrentAccountAsync(new[] {"user.read"}, account)
+                string token = await _msalHelper.GetTokenForCurrentAccountAsync(new[] { "user.read" }, account)
                     .ConfigureAwait(false);
-                DisplayUserInformationFromGraph(token);
```
Comment (bgavrilMS) [inline msal/samples/desktop/SampleApp/MainForm.cs:62]: Why was this removed?  #Closed

File: msal/samples/desktop/SampleApp/MsalAuthHelper.cs [hunk 2]
```diff
@@ -98,5 +112,32 @@ public async Task<string> GetTokenForCurrentAccountAsync(IEnumerable<string> sco
...
+        {
+            AuthenticationResult result = null;
+            Exception exception = null;
+
+            try
+            {
+                result = await Application.AcquireTokenByIntegratedWindowsAuthAsync(scopes, user).ConfigureAwait(false);
+                return result.AccessToken;
+            }
+            catch (MsalServiceException ex)
```
Comment (bgavrilMS) [inline msal/samples/desktop/SampleApp/MsalAuthHelper.cs]: You can just catch all exception, like at line 130 and then show the message box, like at line 137.  #Resolved

File: msal/samples/desktop/SampleApp/MsalAuthHelper.cs [hunk 3]
```diff
@@ -37,54 +38,67 @@ namespace SampleApp
...
-            catch (Exception exc)
+            catch (MsalUiRequiredException)
             {
-                MessageBox.Show(exc.Message, "Sign in failed", MessageBoxButtons.OK, MessageBoxIcon.Error);
+                try
+                {
+                    result = await Application.AcquireTokenAsync(scopes).ConfigureAwait(false);
+                    return result.AccessToken;
+                }
+                catch (MsalServiceException ex)
```
Comment (bgavrilMS) [inline msal/samples/desktop/SampleApp/MsalAuthHelper.cs]: Not needed since you catch all exceptions at line 74 #Resolved

File: msal/src/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 4]
```diff
@@ -277,6 +277,11 @@ public async Task RemoveAsync(IAccount account)
         internal async Task<AuthenticationResult> AcquireTokenSilentCommonAsync(Authority authority,
             IEnumerable<string> scopes, IAccount account, bool forceRefresh, ApiEvent.ApiIds apiId)
         {
+            if (account == null)
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/ClientApplicationBase.cs]: This is part of a different change and should not be here. I think you forgot to rebase all changes to the current dev. Still, this should not be here. #Pending

File: msal/src/Microsoft.Identity.Client/Exceptions/MsalErrorMessage.cs [hunk 5]
```diff
@@ -130,6 +130,8 @@ internal static class MsalErrorMessage
 
         public const string DuplicateQueryParameterTemplate = "Duplicate query parameter '{0}' in extraQueryParameters";
         public const string DeviceCertificateNotFoundTemplate = "Device Certificate was not found for {0}";
+        public const string MsalUiRequiredMessage =
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/Exceptions/MsalErrorMessage.cs]: Different change #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs [hunk 6]
```diff
@@ -85,14 +84,22 @@ private async Task FetchAssertionFromWsTrustAsync()
...
                     userAssertion = new UserAssertion(
                         wsTrustResponse.Token,
                         (wsTrustResponse.TokenType == WsTrustResponse.Saml1Assertion) ? OAuth2GrantType.Saml11Bearer : OAuth2GrantType.Saml20Bearer);
                 }
+                else if (AuthenticationRequestParameters.Authority.AuthorityType == Core.Instance.AuthorityType.Aad)
+                {
+                    // handle grant flow
+                    if (!this.usernamePasswordInput.HasPassword())
+                    {
+                        throw new MsalException(MsalError.PasswordRequiredForManagedUserError);
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs]: We should always throw MsalClientException or MsalServiceException. We expose MsalException as a base class as a convenience for users to catch and treat both at the same time, but we should never throw it.  #Resolved

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 7]
```diff
@@ -511,16 +511,43 @@ public async Task<AuthenticationResult> AcquireTokenByIntegratedWindowsAuthAsync
...
 #if DESKTOP
-        public async Task<AuthenticationResult> AcqjkreTokenByUsernamePassword(IEnumerable<string> scopes, System.Security.SecureString password)
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="scopes"></param>
+        /// <param name="username"></param>
+        /// <param name="password"></param>
+        /// <returns></returns>
+        public async Task<AuthenticationResult> AcquireTokenByUsernamePasswordAsync(IEnumerable<string> scopes, string username, string password)
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: I would propose that we don't expose this overload (at least not on .net desktop where we have SecureString). The reasoning is that we should offer users a single API for achieving U/P, and we should prefer the more secure version of this API (i.e. the one at line 534 with SecureString). @henrik-me - can you think of a user scenario where they cannot use SecureString but need to use string  on net45 ? #Resolved

File: core/src/Exceptions/CoreErrorMessages.cs [hunk 8]
```diff
@@ -83,7 +83,7 @@ internal class CoreErrorMessages
         public const string ParsingWsTrustResponseFailed = "Parsing WS-Trust response failed";
         public const string UnknownUserType = "Unknown User Type";
 
-
-
+        public const string InternalErrorCacheEmptyUsername = "Internal error - trying to remove an ADAL user with an empty username. Possible cache corruption. See https://aka.ms/adal_token_cache_serialization";
```
Comment (bgavrilMS) [inline core/src/Exceptions/CoreErrorMessages.cs]: Merge concflicts ? Are these related to IWA / UP ? #Resolved

File: core/tests/Test.Microsoft.Identity.Core.Unit/Mocks/MockHttpMessageHandler.cs [hunk 9]
```diff
@@ -47,6 +47,8 @@ internal class MockHttpMessageHandler : HttpMessageHandler
 
         public IDictionary<string, string> PostData { get; set; }
 
+        public IDictionary<string, object> PostDataObject { get; set; }
```
Comment (bgavrilMS) [inline core/tests/Test.Microsoft.Identity.Core.Unit/Mocks/MockHttpMessageHandler.cs:50]: Shouldn't this be used  in this class? #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs [hunk 10]
```diff
@@ -122,9 +122,17 @@ private async Task UpdateUsernameAsync()
 
         protected override void SetAdditionalRequestParameters(OAuth2Client client)
         {
-            client.AddBodyParameter(OAuth2Parameter.GrantType, OAuth2GrantType.Password);
-            client.AddBodyParameter(OAuth2Parameter.Username, this.usernamePasswordInput.UserName);
-            client.AddBodyParameter(OAuth2Parameter.Password, new string(this.usernamePasswordInput.PasswordToCharArray()));
+            if (this.userAssertion != null)
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs]: I remember you saying this condition is hit only when the account is managed? If so, could you add a comment to make it clearer? #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs [hunk 11]
```diff
@@ -134,8 +133,13 @@ protected override void SetAdditionalRequestParameters(OAuth2Client client)
                 client.AddBodyParameter(OAuth2Parameter.Password, new string(this.usernamePasswordInput.PasswordToCharArray()));
             }
 
-            // To request id_token in response
-            client.AddBodyParameter(OAuth2Parameter.Scope, OAuth2Value.ScopeOpenId);
+            SortedSet<string> unionScope =
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs]: Why SortedSet instead of a HashSet? Is it for debugging, i.e. to see have the Get params sorted alphabetically in Fiddler? #Resolved

File: core/src/OAuth2/OAuthConstants.cs [hunk 12]
```diff
@@ -102,6 +102,8 @@ internal class OAuth2Value
         public static readonly string[] ReservedScopes = {"openid", "profile", "offline_access"};
```
Comment (bgavrilMS) [inline core/src/OAuth2/OAuthConstants.cs]: You can update this constant to use ScopeOpenId, ScopeOfflineAccess etc. #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs [hunk 13]
```diff
@@ -415,164 +415,165 @@ public async Task AcquireTokenByIntegratedWindowsAuthTest()
...
-            AuthenticationResult result = await app.AcquireTokenByUsernamePasswordAsync(TestConstants.Scope, TestConstants.User.Username, str).ConfigureAwait(false);
-
-            Assert.IsNotNull(result);
-            Assert.AreEqual("some-access-token", result.AccessToken);
-            Assert.IsNotNull(result.Account);
-            Assert.AreEqual(TestConstants.User.Username, result.Account.Username);
-
-            Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
-        }
+        //[TestMethod]
```
Comment (bgavrilMS) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: Lots of dead code, I assume these are uncommented in a subsequent commit? #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs [hunk 14]
```diff
@@ -416,164 +416,172 @@ public async Task AcquireTokenByIntegratedWindowsAuthTest()
...
-        //    AuthenticationResult result = await app.AcquireTokenByUsernamePasswordAsync(TestConstants.Scope, TestConstants.User.Username, str).ConfigureAwait(false);
-
-        //    Assert.IsNotNull(result);
-        //    Assert.AreEqual("some-access-token", result.AccessToken);
-        //    Assert.IsNotNull(result.Account);
-        //    Assert.AreEqual(TestConstants.User.Username, result.Account.Username);
-
-        //    Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
-        //}
+        [TestMethod]
```
Comment (bgavrilMS) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: So they are... #Resolved

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 15]
```diff
@@ -512,28 +512,28 @@ public async Task<AuthenticationResult> AcquireTokenByIntegratedWindowsAuthAsync
...
-        /// <param name="scopes"></param>
-        /// <param name="username"></param>
-        /// <param name="password"></param>
-        /// <returns></returns>
-        public async Task<AuthenticationResult> AcquireTokenByUsernamePasswordAsync(IEnumerable<string> scopes, string username, System.Security.SecureString password)
+        /// <param name="scopes">Scopes requested to access a protected API</param>
+        /// <param name="username">Identifier of the user application requests token on behalf.</param>
+        /// <param name="securePassword">User password.</param>
+        /// <returns>Authentication result containing a token for the requested scopes and account</returns>
+        public async Task<AuthenticationResult> AcquireTokenByUsernamePasswordAsync(IEnumerable<string> scopes, string username, System.Security.SecureString securePassword)
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs:521]: @jmprieur to review the public API comments.

File: msal/tests/Test.MSAL.NET.Unit/IWATests.cs [hunk 16]
```diff
@@ -0,0 +1,623 @@
...
+        {
+            cache.tokenCacheAccessor.AccessTokenCacheDictionary.Clear();
+            cache.tokenCacheAccessor.RefreshTokenCacheDictionary.Clear();
+        }
+
+        [TestMethod]
+        [TestCategory("IWATests")]
+        [DeploymentItem(@"Resources\TestMex.xml", "MsalResource")]
+        [DeploymentItem(@"Resources\WsTrustResponse13.xml", "MsalResource")]
+        public async Task AcquireTokenByIntegratedWindowsAuthTest()
```
Comment (bgavrilMS) [inline msal/tests/Test.MSAL.NET.Unit/IWATests.cs]: Excellent!

File: msal/tests/Test.MSAL.NET.Unit/IWATests.cs [hunk 17]
```diff
@@ -0,0 +1,623 @@
...
+                    Content = new StringContent("{\"ver\":\"1.0\",\"account_type\":\"federated\",\"domain_name\":\"microsoft.com\"," +
+                                                "\"federation_protocol\":\"WSTrust\",\"federation_metadata_url\":" +
+                                                "\"https://msft.sts.microsoft.com/adfs/services/trust/mex\"," +
+                                                "\"federation_active_auth_url\":\"https://msft.sts.microsoft.com/adfs/services/trust/2005/usernamemixed\"" +
+                                                ",\"cloud_instance_name\":\"login.microsoftonline.com\"}")
+                }
+            });
+
+            // MEX
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
```
Comment (bgavrilMS) [inline msal/tests/Test.MSAL.NET.Unit/IWATests.cs]: There seem to be a lot of these duplicated code blocks for setting up mocks. See if you can create some private methods that can be reused.  #Resolved

File: msal/tests/Test.MSAL.NET.Unit/IWATests.cs [hunk 18]
```diff
@@ -0,0 +1,623 @@
...
+using System.Net.Http;
+using System.Security;
+using System.Threading.Tasks;
+using Test.Microsoft.Identity.Core.Unit.Mocks;
+using Test.MSAL.NET.Unit.Mocks;
+
+namespace Test.MSAL.NET.Unit
+{
+    [TestClass]
+    public class IWATests
```
Comment (bgavrilMS) [inline msal/tests/Test.MSAL.NET.Unit/IWATests.cs]: There are username password tests in here as well, not just iwa #Resolved

File: msal/tests/Test.MSAL.NET.Unit/IWATests.cs [hunk 19]
```diff
@@ -0,0 +1,623 @@
...
+
+            // There should be no cached entries.
+            Assert.AreEqual(0, cache.tokenCacheAccessor.AccessTokenCacheDictionary.Count);
+
+            Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
+        }
+
+        [TestMethod]
+        [TestCategory("IWATests")]
+        public async Task ManagedUsernamePasswordAcquireTokenTest()
```
Comment (bgavrilMS) [inline msal/tests/Test.MSAL.NET.Unit/IWATests.cs]: Could we add a test for UsernamePassword where the password in wrong? I think this is a big scenario that should be covered. #Resolved

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
Looks good, some changes requested.

File: msal/src/Microsoft.Identity.Client/Exceptions/MsalErrorMessage.cs [hunk 20]
```diff
@@ -97,7 +97,7 @@ internal static class MsalErrorMessage
         public const string UnexpectedAuthorityValidList = "Unexpected list of valid addresses";
 
         public const string UnknownUser = "Could not identify logged in user";
-        public const string UnsupportedUserType = "Unsupported User Type '{0}'";
+        public const string UnsupportedUserType = "Unsupported User Type '{0}'. Possible cause: the user {1} is not domain joined.";
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Exceptions/MsalErrorMessage.cs]: pls. add an aka.ms link to our wiki pages talking about this error. then we can update our docs to explain how to resolve it and even in the service cases point to external docs. As a starting point I suggest simply pointing to an error page for client exceptions and anotherone for service exceptions. happy to discuss more. Clarification: we should imo also add a link to a general error page, though if the one already linked clearly explains the error scenario then that is naturally better (though the general error information page should explain this as well or point to the right place).  #Resolved

File: msal/tests/dev apps/DesktopTestApp/DesktopTestApp.csproj [hunk 21]
```diff
@@ -42,6 +42,7 @@
     <OutputPath>bin\Appveyor\</OutputPath>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Microsoft.VisualBasic" />
```
Comment (henrik-me) [inline msal/tests/dev apps/DesktopTestApp/DesktopTestApp.csproj]: ? #Resolved

Comment (bgavrilMS) [reply msal/tests/dev apps/DesktopTestApp/DesktopTestApp.csproj]: Yeah, it looks like VB has a simple dialog that can be used to fetch an input box. It helps with testing, although you cannot control the text box to set it to mask the text.  #Resolved

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs (hunk 7 above)
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: I can't think of a scenario right now. but there might be one... 

---
In reply to: [219455652](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219455652) [](ancestors = 219455652)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs [hunk 22]
```diff
@@ -94,8 +94,13 @@ private async Task FetchAssertionFromWsTrustAsync()
...
                 {
-                    throw new MsalException(MsalError.UnknownUserType,
-                        string.Format(CultureInfo.CurrentCulture, MsalErrorMessage.UnsupportedUserType, userRealmResponse.AccountType));
+                    throw new MsalException(
+                        MsalError.UnknownUserType,
+                        string.Format(
+                            CultureInfo.CurrentCulture, 
+                            MsalErrorMessage.UnsupportedUserType, 
+                            userRealmResponse.AccountType, 
+                            this.iwaInput.UserName));
```
Comment (jennyf19) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs]: Is this considered Pii? #Resolved

Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs]: yes, however if only shown to the user it is fine. if we log it... we have to filter it out

---
In reply to: [219635839](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219635839) [](ancestors = 219635839)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs (hunk 6 above)
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs]: Seems like this should be a client exception as Bogdan suggests. Same is in IWarRequest

---
In reply to: [219454654](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219454654) [](ancestors = 219454654)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+using Microsoft.Identity.Core.Helpers;
+using Microsoft.Identity.Core.OAuth2;
+using Microsoft.Identity.Core.WsTrust;
+
+namespace Microsoft.Identity.Client.Internal.Requests
+{
+    /// <summary>
+    /// Handles requests that are non-interactive. Currently MSAL supports Integrated Windows Auth.
+    /// </summary>
+    internal class IWARequest : RequestBase
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs]: >IWARequest [](start = 19, length = 10)

Perhaps this should be spelled out. IWA sounds like some interface... #Resolved

File: msal/tests/dev apps/DesktopTestApp/MainForm.cs [hunk 24]
```diff
@@ -158,6 +161,45 @@ private async void AcquireTokenInteractive_Click(object sender, EventArgs e)
...
+                CreateException(exc);
+            }
+        }
+               
+        private async void acquireTokenByUPButton_Click(object sender, EventArgs e)
+        {
+            ClearResultPageInfo();
+
+            string username = loginHintTextBox.Text; //Can be blank for U/P
+            var password = Microsoft.VisualBasic.Interaction.InputBox("Password?");
```
Comment (henrik-me) [inline msal/tests/dev apps/DesktopTestApp/MainForm.cs]: > Microsoft.VisualBasic.Interaction.InputBox("Password?"); [](start = 26, length = 57)

nit: It is possible to set the text box in the ui to be a password box... then no need to introduce a VB control? #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs (hunk 22 above)
Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs]: ok. we're not logging it. #Resolved

File: msal/samples/desktop/SampleApp/MainForm.cs (hunk 1 above)
Comment (jennyf19) [reply msal/samples/desktop/SampleApp/MainForm.cs:62]: it wasn't working...i don't think it ever has worked.

---
In reply to: [219452300](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219452300) [](ancestors = 219452300)

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs (hunk 7 above)
Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: @henrikm and i discussed this...we will only expose securestring for now.

---
In reply to: [219528465](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219528465) [](ancestors = 219528465,219455652)

File: core/tests/Test.Microsoft.Identity.Core.Unit/Mocks/MockHttpMessageHandler.cs (hunk 9 above)
Comment (jennyf19) [reply core/tests/Test.Microsoft.Identity.Core.Unit/Mocks/MockHttpMessageHandler.cs:50]: this is needed for the mocks, to handle the securestring.

---
In reply to: [219473186](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219473186) [](ancestors = 219473186)

File: msal/tests/Test.MSAL.NET.Unit/IWATests.cs (hunk 17 above)
Comment (jennyf19) [reply msal/tests/Test.MSAL.NET.Unit/IWATests.cs]: yes...thanks!

---
In reply to: [219476190](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219476190) [](ancestors = 219476190)

File: msal/tests/Test.MSAL.NET.Unit/IWATests.cs (hunk 18 above)
Comment (jennyf19) [reply msal/tests/Test.MSAL.NET.Unit/IWATests.cs]: i renamed it, but open to a better name. 

---
In reply to: [219477471](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219477471) [](ancestors = 219477471)

File: msal/samples/desktop/SampleApp/MainForm.cs [hunk 25]
```diff
@@ -52,31 +56,83 @@ public MainForm()
...
+
+        private async Task AcquireTokenAsync()
+        {
+            token = await _msalHelper.GetTokenForCurrentAccountAsync(new[] { "user.read" }, account).ConfigureAwait(false);
+        }
+
+        private async void acquireTokenUsernamePasswordButton_Click(object sender, EventArgs e)
+        {
+            tokenResultBox.Text = string.Format("");
+            string password = tokenResultBox.Text;
```
Comment (henrik-me) [inline msal/samples/desktop/SampleApp/MainForm.cs]: >string password = tokenResultBox.Text; [](start = 12, length = 38)

you should pass the textboxresult to the method, otherwise you are copying the string into memory anyway... here you end up doing it twice #Resolved

Comment (henrik-me) [reply msal/samples/desktop/SampleApp/MainForm.cs]: then...:

var secure = new SecureString();
foreach (char c in textbox1.Text) {
  secure.AppendChar(c);
}

---
In reply to: [219649112](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219649112) [](ancestors = 219649112)

Comment (jennyf19) [reply msal/samples/desktop/SampleApp/MainForm.cs]: defeats the point of the secure string...i fixed this.

---
In reply to: [219649259](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219649259) [](ancestors = 219649259,219649112)

File: msal/tests/dev apps/DesktopTestApp/MainForm.cs (hunk 24 above)
Comment (jennyf19) [reply msal/tests/dev apps/DesktopTestApp/MainForm.cs]: yes. but less fun.

---
In reply to: [219640245](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219640245) [](ancestors = 219640245)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs (hunk 23 above)
Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs]: I spelled it out everywhere.

---
In reply to: [219640027](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219640027) [](ancestors = 219640027)

File: msal/tests/dev apps/DesktopTestApp/DesktopTestApp.csproj (hunk 21 above)
Comment (jennyf19) [reply msal/tests/dev apps/DesktopTestApp/DesktopTestApp.csproj]: i removed the reference.

---
In reply to: [219518522](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219518522) [](ancestors = 219518522)

File: msal/tests/Test.MSAL.NET.Unit/IWATests.cs (hunk 16 above)
Comment (jennyf19) [reply msal/tests/Test.MSAL.NET.Unit/IWATests.cs]: :)

---
In reply to: [219475814](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219475814) [](ancestors = 219475814)

File: core/src/Exceptions/CoreErrorMessages.cs (hunk 8 above)
Comment (jennyf19) [reply core/src/Exceptions/CoreErrorMessages.cs]: yes

---
In reply to: [219456012](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219456012) [](ancestors = 219456012)

File: msal/src/Microsoft.Identity.Client/ClientApplicationBase.cs (hunk 4 above)
Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/ClientApplicationBase.cs]: i had to add this for testing, but I did rebase w/current dev, but maybe because i added it earlier (for testing), the changes w/dev were not accounted for?

---
In reply to: [219454270](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219454270) [](ancestors = 219454270)

File: msal/src/Microsoft.Identity.Client/MigrationAid/MigrationAid.cs [hunk 26]
```diff
@@ -145,4 +146,26 @@ public partial class TokenCacheNotificationArgs
         [Obsolete("Use Account instead (See https://aka.ms/msal-net-2-released)", true)]
         public IUser User { get { throw new NotImplementedException(); } }
     }
+
+    /// <Summary>
+    /// Abstract class containing common API methods and properties. 
+    /// For details see https://aka.ms/msal-net-client-applications
+    /// </Summary>
+    public partial class PublicClientApplication
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/MigrationAid/MigrationAid.cs:154]: >PublicClientApplication [](start = 25, length = 23)

We never had this method in MSAL, why do we want to obselete an inexistent method? #Resolved

Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/MigrationAid/MigrationAid.cs:154]: This is to help those upgrading from ADAL trying to use the string version of the pwd. Telling them to use the SecureString intead.
